byterun/intern.c

```diff
@@ -718,15 +720,25 @@ value caml_input_val(struct channel *chan)
   }
   /* Initialize global state */
   intern_init(block, block);
-  intern_alloc(h.whsize, h.num_objects);
+  intern_alloc(h.whsize, h.num_objects, outside_heap);
   /* Fill it in */
   intern_rec(&res);
-  intern_add_to_heap(h.whsize);
+  if (!outside_heap) {
+    intern_add_to_heap(h.whsize);
+  } else {
+    intern_extra_block = NULL;
+    intern_block = 0;
+  }
```

Comment:
> We clearly cannot free the block there since it contains the demarshalled data.

Oh, indeed. I was the one misunderstanding.

> We could change the comment in memory.c to reflect that this is safe.

Yes, that could be nice (documented calling protocols should be respected or fixed). Another approach would be to have a dummy function `caml_{release,disown}_from_heap` that would be a no-op but clarify the usage protocol. I'm thinking of people willing to change/instrument the allocation primitives (in the style of #71) may be interested in the ability to override functions to keep track of what is owned by the OCaml heap and what is owned outside it.